drivers/include/matrix_keypad.h

```diff
@@ -0,0 +1,218 @@
+ * depending on the number of columns.
+ */
+#if CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 8
+typedef uint8_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 16
+typedef uint16_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 32
+typedef uint32_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 64
+typedef uint64_t matrix_keypad_state_row_t;
+#else
+#error Too many columns on matrix keypad.
+#endif
+
+/**
+ * @brief   Device initialization parameters
+ */
+typedef struct {
+    /**
+     * @brief GPIO pin array for the rows. GPIO_UNDEF is allowed and scanning
+     * is skipped for these.
+     */
+    gpio_t rows[CONFIG_MATRIX_KEYPAD_NUM_ROWS];
+    /**
+     * @brief GPIO pin array for the columns. GPIO_UNDEF is allowed and scanning
+     * is skipped for these.
+     */
+    gpio_t columns[CONFIG_MATRIX_KEYPAD_NUM_COLUMNS];
+    /**
+     * @brief Delay in microseconds between configuring the row gpio and reading
+     * out the column. Can be zero to skip the delay.
+     */
+    uint32_t row2col_delay;
+} matrix_keypad_params_t;
+
+/**
+ * @brief   Callback for key state changes
+ *
+ * @param   arg     callback context
+ * @param   row     Row that changed
+ * @param   column  Column that changed
+ * @param   state   New state of the key, 1 = pressed, 0 = released
+ */
+typedef void (*matrix_keypad_cb_t)(void *arg, size_t row, size_t column, bool state);
+
+/**
+ * @brief   Device descriptor for the driver
+ */
+typedef struct {
+    /**
+     * @brief Device initialization parameters
+     */
+    matrix_keypad_params_t params;
+
+    /**
+     * @brief Debounce history
+     */
+    uint8_t debounce[CONFIG_MATRIX_KEYPAD_NUM_ROWS][CONFIG_MATRIX_KEYPAD_NUM_COLUMNS];
+
+    /**
+     * @brief Current button state
+     */
+    matrix_keypad_state_row_t state[CONFIG_MATRIX_KEYPAD_NUM_ROWS];
+
+    /**
+     * @brief callback context
+     */
+    void *arg;
+
+    /**
+     * @brief Callback called when a key changes state
+     */
+    matrix_keypad_cb_t callback;
+} matrix_keypad_t;
+
+/**
+ * @brief   Initialize the given device
+ *
+ * @param[inout] dev        Device descriptor of the driver
+ * @param[in]    params     Initialization parameters
+ * @param[in]    callback   Callback to call on state changes
+ * @param[in]    arg        Context argument for the callback
+ *
+ * @return                  0 on success
+ */
+int matrix_keypad_init(matrix_keypad_t *dev,
+                       const matrix_keypad_params_t *params,
+                       matrix_keypad_cb_t callback,
+                       void *arg);
+
+/**
+ * @brief   Scan through the keypad matrix
+ *
+ * This updates the state of the device descriptor, calling
+ * @ref matrix_keypad_cb_t when a key press or release has been detected
+ *
+ * @param[inout] dev        Device descriptor of the driver
+ */
+unsigned matrix_keypad_scan(matrix_keypad_t *dev);
```

Comment:
What is the return value?

Reply:
Number of keys changed, changed to `size_t` and added docs.